pkg/autodiscovery/listeners/snmp.go

```diff
@@ -0,0 +1,356 @@
+	for _, config := range l.config.Configs {
+		ipAddr, ipNet, err := net.ParseCIDR(config.Network)
+		if err != nil {
+			log.Errorf("Couldn't parse SNMP network: %s", err)
+			continue
+		}
+
+		defaultParams, err := config.BuildSNMPParams()
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+
+		startingIP := ipAddr.Mask(ipNet.Mask)
+		currentIP := make(net.IP, len(startingIP))
+		copy(currentIP, startingIP)
+
+		configHash := config.Digest(config.Network)
+		cacheKey := fmt.Sprintf("snmp:%s", configHash)
+		adIdentifier := cacheKey
+		l.loadCache(config, adIdentifier, cacheKey)
+
+		subnet := snmpSubnet{
+			adIdentifier:  adIdentifier,
+			config:        config,
+			defaultParams: defaultParams,
+			startingIP:    startingIP,
+			currentIP:     currentIP,
+			network:       *ipNet,
+			cacheKey:      cacheKey,
+		}
+		subnets = append(subnets, subnet)
+	}
+
+	if l.config.Workers == 0 {
+		l.config.Workers = 2
+	}
+
+	if l.config.DiscoveryInterval == 0 {
+		l.config.DiscoveryInterval = 3600
+	}
+
+	jobs := make(chan snmpSubnet)
+	for w := 0; w < l.config.Workers; w++ {
+		go worker(l, jobs)
+	}
+
+	subnetsDone := 0
+	discoveryTicker := time.NewTicker(time.Duration(l.config.DiscoveryInterval) * time.Second)
+
+	for {
+		for _, subnet := range subnets {
+			if subnet.network.Contains(subnet.currentIP) {
+				ignored := false
+				currentString := subnet.currentIP.String()
+				for _, ip := range subnet.config.IgnoredIPAddresses {
+					if ip == currentString {
+						ignored = true
+						break
+					}
+				}
+				if ignored {
+					incrementIP(subnet.currentIP)
+					continue
+				}
+				jobSubnet := subnet
+				jobSubnet.currentIP = make(net.IP, len(subnet.currentIP))
+				copy(jobSubnet.currentIP, subnet.currentIP)
+				jobs <- jobSubnet
+				incrementIP(subnet.currentIP)
+			} else {
+				// XXX write the cache a bit more often
+				l.writeCache(subnet.cacheKey, subnet.adIdentifier)
+				subnetsDone++
+			}
+		}
+
+		if subnetsDone == len(subnets) {
+			for _, subnet := range subnets {
+				copy(subnet.currentIP, subnet.startingIP)
+			}
+			subnetsDone = 0
+
+			select {
+			case <-l.stop:
+				return
+			case <-discoveryTicker.C:
+			}
+		} else {
+			select {
+			case <-l.stop:
+				return
+			default:
+			}
+		}
+	}
+}
+
+func (l *SNMPListener) createService(deviceIP string, adIdentifier string, entityID string) {
+	l.Lock()
+	defer l.Unlock()
+	if _, present := l.services[entityID]; present {
+		return
+	}
+	svc := &SNMPService{
+		adIdentifier: adIdentifier,
+		entityID:     entityID,
+		deviceIP:     deviceIP,
+		creationTime: integration.Before,
+	}
+	l.services[entityID] = svc
+	l.devices[adIdentifier] = append(l.devices[adIdentifier], deviceIP)
+	l.newService <- svc
+}
+
+func (l *SNMPListener) deleteService(entityID string) {
+	l.Lock()
+	defer l.Unlock()
+	// XXX don't delete on first failure
+	svc, present := l.services[entityID]
+	if present {
+		l.delService <- svc
+		delete(l.services, entityID)
+	}
```

Comment:
Could this be shortened to the following?

```suggestion
	if svc, present := l.services[entityID]; present {
		l.delService <- svc
		delete(l.services, entityID)
	}
```